pkg/ext-proc/scheduling/filter.go

```diff
@@ -0,0 +1,127 @@
+package scheduling
+
+import (
+	"fmt"
+	"math"
+
+	klog "k8s.io/klog/v2"
+
+	"ext-proc/backend"
+)
+
+type Filter interface {
+	Name() string
+	Filter(b *LLMRequest, pods []*backend.PodMetrics) ([]*backend.PodMetrics, error)
+}
+
+// filter applies current filterFunc, and then recursively applies next filters depending success or
+// failure of the current filterFunc.
+// It can be used to construct a flow chart algorithm.
+type filter struct {
+	name   string
+	filter filterFunc
+	// nextOnSuccess filter will be applied after successfully applying the current filter.
+	// The filtered results will be passed to the next filter.
+	nextOnSuccess *filter
+	// nextOnFailure filter will be applied if current filter fails.
+	// The original input will be passed to the next filter.
+	nextOnFailure *filter
+	// nextOnSuccessOrFailure is a convenience field to configure the next filter regardless of the
+	// success or failure of the current filter.
+	nextOnSuccessOrFailure *filter
+}
+
+func (f *filter) Name() string {
+	if f == nil {
+		return "nil"
+	}
+	return f.name
+}
+
+func (f *filter) Filter(b *LLMRequest, pods []*backend.PodMetrics) ([]*backend.PodMetrics, error) {
+	if f == nil {
+		klog.V(2).Infof("Running nil filter, returning all input pods by default")
+		return pods, nil
+	}
+	klog.V(2).Infof("Running filter %q on request %v with %v pods", f.name, b, len(pods))
+
+	filtered, err := f.filter(b, pods)
+
+	next := f.nextOnSuccessOrFailure
+	if err == nil {
+		klog.V(2).Infof("onSuccess %v -> %v, filtered: %v", f.name, next.Name(), len(filtered))
+		if f.nextOnSuccess != nil {
+			next = f.nextOnSuccess
+		}
+		// On success, pass the filtered result to the next filter.
+		return next.Filter(b, filtered)
+	} else {
+		klog.V(2).Infof("onFailure %v -> %v", f.name, next.Name())
+		if f.nextOnFailure != nil {
+			next = f.nextOnFailure
+		}
+		// On failure, pass the initial set of pods to the next filter.
+		return next.Filter(b, pods)
+	}
+}
```

Comment:
I think we could simplify as this. IMO, it simplify readability by removing the `else` statement:
```suggestion
        next := f.nextOnSuccessOrFailure
        if err == nil {
	     klog.V(2).Infof("onSuccess %v -> %v, filtered: %v", f.name, next.Name(), len(filtered))
	        if f.nextOnSuccess != nil {
		        next = f.nextOnSuccess
	        }
	        // Pass the filtered result on success.
	        return next.Filter(b, filtered)
        } 

        klog.V(2).Infof("onFailure %v -> %v", f.name, next.Name())
        if f.nextOnFailure != nil {
	        next = f.nextOnFailure
        }
        // Pass the initial set of pods on failure.
        return next.Filter(b, pods)
}
```
PS: I don't see how to properly format on code review. I'm letting you update on your side if it looks better to you.

Reply:
done